### ISSUE TITLE
EF CORE 5.0 ThenInclude OrderBy, uses PK before OrderBy field, rendering it useless

### ISSUE THREAD

**Issue by waywardcode** [aspnet/EntityFrameworkCore#26343]
Has there been any progress with this? Ordering navigation properties still does not translate into the expected ordering in EFC5.0.5

```cs
var form = _db.Forms
  .Include(x => x.Sections)
  .ThenInclude(x => x.Rows.OrderBy(y => y.Order)) //Nope!
  .ThenInclude(x => x.Fields.OrderBy(y => y.Order)) //Nope here too!
  .ThenInclude(x => x.FieldType)
  .AsSplitQuery()
  .SingleOrDefaultAsync(x =>
      x.Id == formId)
```

If there is no intention of ever fixing this, it might be best to show a warning when an attempt to order a navigation property is made to avoid confusion.

_Originally posted by @Bitz in https://github.com/dotnet/efcore/issues/9067#issuecomment-843671773_

Example output on ORDERBY inside ThenInclude from EF

ORDER BY [a].[AreaId], [a].[NavBlockSortOrder], [a].[NavBlockId], [n].[AreaId], [n0].[NavBlockId], [t0].[NavGroupId], [t0].[NavBlockId], [t0].[SortOrder]

- notice the last three, switching [t0].[NavGropuId] with [t0].[SortOrder] works as expected,  but not as shown above

Any solution to this, hidden unknown dark magic or otherwise? :)

**Comment by Bitz** [comment]
> This issue is lacking enough information for us to effectively reproduce. Please post a runnable project/solution or complete code listing that demonstrates the behavior you are seeing.

Hey,

This is essentially a duplicate of #9067

Thanks

**Comment by smitpatel** [comment]
This is not duplicate of #9067 using OrderBy inside Include, tries to populate collection in the order specified inside collection, support for which was added in #1833 We have several test verifying that it works correctly. At the same time it only works if the collection type has a order preserving behavior, like List, but something like HashSet may not preserve the order in which items are added.

#9067 tracks ordered collection which is about ability to preserve the index of item in collection while saving them, so when querying them the collection will have items in same order (this is not the order specified by OrderBy but rather ordering in which user initially populated collection.)

The question still remains, we need runnable repro code to see what you are seeing otherwise we cannot investigate what is going wrong here.

**Comment by waywardcode** [comment]
I don't have way to share my whole repo publically, is there a way I can show you without showing my client's entire project publically?

Here is the offending code - I have tried many, many, combinations (a Weeks worth!!) - so this is what my latest looks like which starts off with a Many-to-Many. I've tried it with one side of the MTM navigation too. i.e. NavArea or NavBlock as starting point. All the same results, include/theninclude deeper in the chain will not play well:

````
private async Task<List<NavAreaDTO>> NavAreasWithNavBlocks()
        {
            var mtmlist = await _context.AreaToBlockManyToMany
                .AsNoTracking()
                .OrderBy(ot => ot.AreaId).ThenBy(o => o.NavBlockSortOrder) // this part is working
                .Include(a => a.NavArea)
                .Include(b => b.NavBlock)
                .ThenInclude(g => g.NavGroups.OrderBy(sg => sg.SortOrder)) // This is broken
                .ThenInclude(i => i.NavItems.OrderBy(si => si.Order)) // This is broken               
                .ToListAsync();
           
           // sort child mtm based on property in mtm join table
            var navAreas = await _context.NavAreas.AsNoTracking().ToListAsync();
            var t = new List<NavAreaDTO>();
            foreach (var area in navAreas)
            {
                t.Add(new NavAreaDTO
                {
                    AreaId = area.AreaId,
                    AreaSortOrder = area.AreaSortOrder,
                    AreaName = area.AreaName,
                    NavBlocks = new List<NavBlock>()
                });
            }          
              ``
            foreach (var i in mtmlist)
            {
                var index = t.FindIndex(y => y.AreaId == i.AreaId); 
                t[index].NavBlocks.Add(i.NavBlock);
            }
            return t;
        }
````

Please let me know if need anything else. I really won't be able to make a whole repo just to show that this is broken. Please read my prior post, you can ignore the references to #9067 and not get hung up on semantics. Two people (me and Biz) are telling you the same thing, I explained where to look in my previous post. Not sure how much more I can do to help the MS EF team. Also, I can do a Teams call if you are really interested in fixing this issue. Also, if I am doing something wrong or expecting too much from EF I am open to a plain "it won't work"

Thanks Again, for all your efforts!

**Comment by roji** [comment]
> After some looking around and thanks for your code sample, I figured it out. I was using Split Queries in Startup.cs as an option .UseQuerySplittingBehavior

There is no difference between specifying split query on a specific query (with `AsSplitQuery()`) or globally in your model (with `UseQuerySplittingBehavior` on your `UseSqlServer`). Both do the exact same thing - the latter simply configures split query as a general default.

Looking at the tweaked code sample, thing still look OK for me. Here's the last of the split queries:

```sql
SELECT [t0].[NavItemId], [t0].[ImageUrl], [t0].[ItemName], [t0].[ItemSortOrder], [t0].[Link], [t0].[NavGroupId], [a].[NavBlockId], [a].[AreaId], [n].[AreaId], [n0].[NavBlockId], [t].[NavGroupId]
FROM [AreaToBlockManyToMany] AS [a]
INNER JOIN [NavAreas] AS [n] ON [a].[AreaId] = [n].[AreaId]
INNER JOIN [NavBlocks] AS [n0] ON [a].[NavBlockId] = [n0].[NavBlockId]
INNER JOIN (
    SELECT [n1].[NavGroupId], [n1].[GroupSortOrder], [n1].[NavBlockId]
    FROM [NavGroups] AS [n1]
) AS [t] ON [n0].[NavBlockId] = [t].[NavBlockId]
INNER JOIN (
    SELECT [n2].[NavItemId], [n2].[ImageUrl], [n2].[ItemName], [n2].[ItemSortOrder], [n2].[Link], [n2].[NavGroupId]
    FROM [NavItems] AS [n2]
) AS [t0] ON [t].[NavGroupId] = [t0].[NavGroupId]
ORDER BY [a].[AreaId], [a].[NavBlockSortOrder], [a].[NavBlockId], [n].[AreaId], [n0].[NavBlockId], [t].[GroupSortOrder], [t].[NavGroupId], [t0].[ItemSortOrder]
```

This query loads the NavItems (aliased to `[t0]`). The ordering is by the AreaToBlockManyToMany (AreaId, NavBlockSortOrder), then by the NavBlock (NavBlockID), then by the NavGroups (GroupSortOrder, NavGroupId), and finally by the NavItem (ItemSortOrder).

I think the missing piece here may be that NavItems are returned sorted by ItemSortOrder, but **within** their specific NavItem; EF Core doesn't load just a list of NavItems without context - it populates them as lists within their NavGroups.

**Comment by smitpatel** [comment]
Issue only happens when having 2 ordered includes chained through ThenInclude. The middle split query has incorrect ordering of OrderBy terms. Need to see what is generated SQL in current bits.
We are unlikely to fix this in 5.0 release.
